### PR TITLE
Fix hotstuff integration

### PR DESCRIPTION
**Pull request by zhangchiqing**
Fix https://github.com/dapperlabs/flow-go/issues/6136

This PR fixes the flaky integration tests, [see details of the reasoning.](https://github.com/dapperlabs/flow-go/issues/6136)

### CODE REVIEW

File: consensus/hotstuff/integration/connect_test.go [hunk 1]
```diff
@@ -63,14 +61,7 @@ func Connect(instances []*Instance) {
 						return nil
 					}
 
-					// put the proposal header into the receivers map
-					// TODO: we should not store the header directly without checking if parent
-					// exists, this has been causing flakeiness.
-					// if parent is missing we should store the block in pending blocks
-					receiver.headers.Store(header.ID(), header)
```
Comment (zhangchiqing) [inline consensus/hotstuff/integration/connect_test.go:70]: This was the root cause of the flakiness

File: consensus/hotstuff/integration/instance_test.go [hunk 2]
```diff
@@ -434,3 +448,23 @@ func (in *Instance) Run() error {
...
+
+func (in *Instance) ProcessBlock(proposal *model.Proposal) {
+	in.updatingBlocks.RLock()
+	_, parentExists := in.headers[proposal.Block.QC.BlockID]
+	defer in.updatingBlocks.RUnlock()
+
+	if parentExists {
+		next := proposal
+		for next != nil {
+			in.headers[next.Block.BlockID] = model.ProposalToFlow(next)
```
Comment (jordanschalm) [inline consensus/hotstuff/integration/instance_test.go:460]: ⚠️  We are writing to `in.headers` but have only acquired the read lock of `updatingBlocks`.

File: consensus/hotstuff/integration/instance_test.go [hunk 3]
```diff
@@ -154,23 +159,28 @@ func NewInstance(t require.TestingT, options ...Option) *Instance {
...
 				ChainID:     "chain",
 				ParentID:    parentID,
-				Height:      parent.(*flow.Header).Height + 1,
+				Height:      parent.Height + 1,
 				PayloadHash: unittest.IdentifierFixture(),
 				Timestamp:   time.Now().UTC(),
 			}
 			require.NoError(t, setter(header))
-			in.headers.Store(header.ID(), header)
+			// in.headers[header.ID()] = header
```
Comment (jordanschalm) [inline consensus/hotstuff/integration/instance_test.go:177]: can we remove this? I think the header is processed locally and stored when we broadcast the proposal

File: consensus/hotstuff/integration/instance_test.go [hunk 4]
```diff
@@ -36,6 +36,7 @@ import (
 )
 
 type Instance struct {
+	log zerolog.Logger
```
Comment (durkmurder) [inline consensus/hotstuff/integration/instance_test.go:39]: where do we actually use this?

File: consensus/hotstuff/integration/instance_test.go [hunk 5]
```diff
@@ -245,8 +257,7 @@ func NewInstance(t require.TestingT, options ...Option) *Instance {
 			proposal := model.ProposalFromFlow(header, parent.View)
 
 			// store locally and loop back to engine for processing
-			in.headers.Store(header.ID(), header)
-			in.queue <- proposal
+			in.ProcessBlock(proposal)
```
Comment (durkmurder) [inline consensus/hotstuff/integration/instance_test.go:260]: I believe compliance engine delivers this using a separate goroutine, should we do the same?

**Review by durkmurder — APPROVED** [review]
Looks good to me, thanks for finding root cause of this.

**Review by jordanschalm — APPROVED** [review]

**Comment by zhangchiqing** [comment]
bors merge

**Comment by zhangchiqing** [comment]
bors merge

**Pull request merged**